src/domain/entities/Indicator.ts

```
@@ -275,38 +280,95 @@ export class Indicator extends Struct<IndicatorAttrs>() {
         return [...idsInNumerator, ...idsInDenominator, ...dataElementCode];
     }
 
-    getCompanionCodesFromRules(): Code[] {
+    getAllCompanionCodesFromRules(): Code[] {
         const outcomeCodes = this.getCompanionCodesByType("outcome");
         const outputCodes = this.getCompanionCodesByType("output");
-        return outcomeCodes.concat(outputCodes);
+        return _([...outcomeCodes.values(), ...outputCodes.values()])
+            .flatten()
+            .uniq()
+            .value();
     }
 
-    getCompanionCodesByType(type: "output" | "outcome"): Code[] {
+    getCompanionCodesScope(): HashMap<Code, IndicatorCompanionScope[]> {
+        const outcomeCodes = this.getCompanionCodesByType("outcome");
+        const outputCodes = this.getCompanionCodesByType("output");
+        return _([...outcomeCodes.keys(), ...outputCodes.keys()])
+            .uniq()
+            .flatMap(scope => {
+                return _([...(outcomeCodes.get(scope) ?? []), ...(outputCodes.get(scope) ?? [])])
+                    .uniq()
+                    .map(code => [code, scope] as [Code, IndicatorCompanionScope]);
```

We tried to avoid explicit cast, Are we doing this to avoid the undefined error in the groupFromMap?

If that's the reason I think we can throw an error because that means we have some invalid data, right? Maybe I'm missing something.

```ts
.groupFromMap(([code, scope]) => {
    if (!code || !scope) throw new Error("Invalid code or scope");
    return [code, scope];
});
```

---

I think it is safe to remove. Sorry, I was trying different things at the time and I may have forgotten to clean this up. Thank you!